core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerClient.scala

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2015-2016 IBM Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package whisk.core.containerpool.docker
+
+import scala.util.Try
+import java.io.FileNotFoundException
+import java.nio.file.Files
+import java.nio.file.Paths
+import scala.concurrent.Future
+import scala.concurrent.ExecutionContext
+import whisk.common.Logging
+import whisk.common.TransactionId
+import whisk.common.LoggingMarkers
+import akka.event.Logging.ErrorLevel
+import scala.util.Failure
+import scala.util.Success
+
+/**
+ * Serves as interface to the docker CLI tool.
+ *
+ * Be cautious with the ExecutionContext passed to this, as the
+ * calls to the CLI are blocking.
+ *
+ * You only need one instance (and you shouldn't get more).
+ */
+class DockerClient(dockerHost: Option[String] = None)(executionContext: ExecutionContext)(implicit log: Logging)
+    extends DockerApi with ProcessRunner {
+    implicit private val ec = executionContext
+
+    // Determines how to run docker. Failure to find a Docker binary implies
+    // a failure to initialize this instance of DockerClient.
+    protected val dockerCmd: Seq[String] = {
+        val alternatives = List("/usr/bin/docker", "/usr/local/bin/docker")
+
+        val dockerBin = Try {
+            alternatives.find(a => Files.isExecutable(Paths.get(a))).get
+        } getOrElse {
+            throw new FileNotFoundException(s"Couldn't locate docker binary (tried: ${alternatives.mkString(", ")}).")
+        }
+
+        val host = dockerHost.map(host => Seq("--host", s"tcp://$host")).getOrElse(Seq.empty[String])
+        Seq(dockerBin) ++ host
+    }
+
+    def run(image: String, args: Seq[String] = Seq.empty[String])(implicit transid: TransactionId): Future[ContainerId] =
+        runCmd((Seq("run", "-d") ++ args ++ Seq(image)): _*).map(ContainerId.apply)
+
+    def inspectIPAddress(id: ContainerId)(implicit transid: TransactionId): Future[ContainerIp] =
```

Comment:
The upcoming DockerContainer code will use `docker inspect` command as fallback if it fails to determine the container's IP address from the file system.